src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,620 @@
+            CancellationToken cancellationToken)
+        {
+            return Task.FromResult(UpdateCompletionList(session, data, cancellationToken));
+        }
+
+        public EditorCompletion.FilteredCompletionModel UpdateCompletionList(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            bool useSuggestionMode = session.Properties.GetProperty<bool>(CompletionItemSource.UseSuggestionMode);
+
+            var filterText = session.ApplicableToSpan.GetText(data.Snapshot);
+
+            // Check if the user is typing a number. If so, only proceed if it's a number
+            // directly after a <dot>. That's because it is actually reasonable for completion
+            // to be brought up after a <dot> and for the user to want to filter completion
+            // items based on a number that exists in the name of the item. However, when
+            // we are not after a dot (i.e. we're being brought up after <space> is typed)
+            // then we don't want to filter things. Consider the user writing:
+            //
+            //      dim i =<space>
+            //
+            // We'll bring up the completion list here (as VB has completion on <space>).
+            // If the user then types '3', we don't want to match against Int32.
+            if (filterText.Length > 0 && char.IsNumber(filterText[0]))
+            {
+                if (!IsAfterDot(data.Snapshot, session.ApplicableToSpan))
+                {
+                    session.Dismiss();
+                    return new EditorCompletion.FilteredCompletionModel(ImmutableArray<EditorCompletion.CompletionItemWithHighlight>.Empty, 0);
+                }
+            }
+
+            // We need to filter if a non-empty strict subset of filters are selected
+            var selectedFilters = data.SelectedFilters.Where(f => f.IsSelected).Select(f => f.Filter).ToImmutableArray();
+            var needToFilter = selectedFilters.Length > 0 && selectedFilters.Length < data.SelectedFilters.Length;
+
+            var initialListOfItemsToBeIncluded = new List<FilterResult>();
+            foreach (var item in data.InitialSortedList)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (needToFilter && ShouldBeFilteredOutOfCompletionList(item, selectedFilters))
+                {
+                    continue;
+                }
+
+                if (MatchesFilterText(item.FilterText, item.DisplayText, GetMatchPriority(item), filterText, data.Trigger.Reason))
+                {
+                    initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: true));
+                }
+                else
+                {
+                    if (data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Deletion ||
```

Comment:
Can we do this earlier, so to avoid actually doing the filtering if we're going to add everyting? I guess it's not obvious to me what "matchedFilterText" gets used for here...

Reply:
It seems that MatchFilterText should have another priority over the reasons comparison: if there is a Match, we set matchedFilterText: true even if it is the deletion. So, we cannot swap condition checks.

---
In reply to: [221394413](https://github.com/dotnet/roslyn/pull/29016#discussion_r221394413) [](ancestors = 221394413)

Reply:
ok.  i found the original code, but things like comments have been stripped.  was there a reason for doing that?

By moving the code to a new location, changing the structure of it, and removing comments, it is *really* hard to ensure that this is doing the right thing. I wish you hadn't done that :(